### PR TITLE
fix: validation erros now send 400 status code

### PR DESCRIPTION
**Pull request by agusaldasoro**

### CODE REVIEW

File: content/src/controller/Controller.ts [hunk 1]
```diff
@@ -119,76 +119,78 @@ export class Controller {
...
-      res.send({
-        creationTimestamp: creationTimestamp
-      })
-    } catch (error) {
-      Controller.LOGGER.warn(`Returning error '${error.message}'`)
-      res.status(500).send(error.message) // TODO: Improve and return 400 if necessary
-    } finally {
-      await this.deleteUploadedFiles(deployFiles)
+    const deployFiles: ContentFile[] = await this.readFiles(files)
+    const creationTimestamp: Timestamp | ErrorList = await this.service.deployLocalLegacy(
```
Comment (nchamo) [inline content/src/controller/Controller.ts:132]: Also, since we are not sure what is returned, we should probably change the name `creationTimestamp`

File: content/src/controller/Controller.ts [hunk 2]
```diff
@@ -119,76 +119,78 @@ export class Controller {
...
+      deployFiles,
+      entityId,
+      auditInfo
+    )
+
+    if (typeof creationTimestamp === 'number') {
+      res.send({ creationTimestamp: creationTimestamp })
+    } else {
+      Controller.LOGGER.warn(`Returning error '${creationTimestamp.join()}'`)
```
Comment (nchamo) [inline content/src/controller/Controller.ts:141]: We used to do `.join('\n')`, let's try to avoid changing it

File: content/src/controller/Controller.ts [hunk 3]
```diff
@@ -119,76 +119,78 @@ export class Controller {
...
-      if (!authChain && ethAddress && signature) {
-        authChain = Authenticator.createSimpleAuthChain(entityId, ethAddress, signature)
-      }
-
-      if (files instanceof Array) {
-        deployFiles = await Promise.all(files.map((f) => this.readFile(f.fieldname, f.path)))
-      }
+    const deployFiles: ContentFile[] = await this.readFiles(files)
+    const auditInfo: LocalDeploymentAuditInfo = this.getAuditInfo(authChain, ethAddress, signature, entityId)
+    let creationTimestamp: Timestamp | ErrorList = []
```
Comment (nchamo) [inline content/src/controller/Controller.ts:161]: Same as before with name, try, and join

File: content/src/controller/Controller.ts [hunk 4]
```diff
@@ -119,76 +119,78 @@ export class Controller {
...
     }
+    await this.deleteUploadedFiles(deployFiles)
   }
 
-  private async readFile(name: string, path: string): Promise<ContentFile> {
-    return {
-      name,
-      path,
-      content: await fs.promises.readFile(path)
+  private getAuditInfo(authChain: AuthLink[], ethAddress: string, signature: string, entityId: string) {
```
Comment (nchamo) [inline content/src/controller/Controller.ts:177]: I think `buildAuditInfo` would be a more descriptive name

File: content/src/service/Service.ts [hunk 5]
```diff
@@ -32,20 +32,20 @@ export interface MetaverseContentService {
     auditInfo: LocalDeploymentAuditInfo,
     origin: string,
     repository?: RepositoryTask | Repository
-  ): Promise<Timestamp>
+  ): Promise<Timestamp | ErrorList>
```
Comment (nchamo) [inline content/src/service/Service.ts:35]: We should probably create a type for `Timestamp | ErrorList`. Something like `DeploymentResult` maybe?

File: content/src/service/Service.ts [hunk 6]
```diff
@@ -93,3 +101,5 @@ export type DeploymentEvent = {
 }
 
 export type DeploymentListener = (deployment: DeploymentEvent) => void | Promise<void>
+
+export type ErrorList = string[]
```
Comment (nchamo) [inline content/src/service/Service.ts:105]: As I said before, something like `InvalidDeploymentResult` would be a better name, and we should probably make this an object like:
```
{ errors: string[] }
```

That way, we can add more information in the future if we need to, without having to make changes

File: content/test/integration/syncronization/error-handling.spec.ts [hunk 7]
```diff
@@ -136,7 +136,7 @@ describe('End 2 end - Error handling', () => {
     removeCauseOfFailure?: () => Promise<void>
   ) {
     // Start servers
```
Comment (nchamo) [inline content/test/integration/syncronization/error-handling.spec.ts:138]:
```suggestion
    // Start server 1
```

File: content/test/integration/syncronization/error-handling.spec.ts [hunk 8]
```diff
@@ -136,7 +136,7 @@ describe('End 2 end - Error handling', () => {
...
-    await Promise.all([server1.start(), server2.start()])
+    await Promise.all([server1.start()])
```
Comment (nchamo) [inline content/test/integration/syncronization/error-handling.spec.ts:139]:
```suggestion
    await server1.start()
```

File: content/test/integration/syncronization/error-handling.spec.ts [hunk 9]
```diff
@@ -152,6 +152,7 @@ describe('End 2 end - Error handling', () => {
     // Cause failure
     await causeOfFailure(entityBeingDeployed)
 
+    await Promise.all([server2.start()])
```
Comment (nchamo) [inline content/test/integration/syncronization/error-handling.spec.ts:155]:
```suggestion
    // Start server 2
    await server2.start()
    
```

File: content/test/integration/syncronization/cluster-synchronization.spec.ts [hunk 10]
```diff
@@ -41,9 +41,10 @@ describe('End 2 end synchronization tests', function () {
     await assertDeploymentsAreReported(server2)
 
     // Deploy the entity to server 1
-    const deploymentTimestamp: Timestamp = await server1.deploy(deployData)
-    const deploymentEvent = buildEvent(entityBeingDeployed, server1, deploymentTimestamp)
-    const deployment = buildDeployment(deployData, entityBeingDeployed, server1, deploymentTimestamp)
+    const deploymentTimestamp: Timestamp | Error = await server1.deploy(deployData)
```
Comment (nchamo) [inline content/test/integration/syncronization/cluster-synchronization.spec.ts:44]: Why `Timestamp | Error`? I'm not sure why this change was necessary

File: content/test/integration/E2ETestUtils.ts [hunk 11]
```diff
@@ -122,8 +122,8 @@ export function awaitUntil(
 export async function deployEntitiesCombo(
   service: MetaverseContentService,
   ...entitiesCombo: EntityCombo[]
-): Promise<Timestamp> {
-  let timestamp: Timestamp = 0
+): Promise<Timestamp | ErrorList> {
+  let timestamp: Timestamp | ErrorList = []
```
Comment (nchamo) [inline content/test/integration/E2ETestUtils.ts:126]: I'm curious, why the change to an empty list by default?

File: content/src/service/synchronization/EventDeployer.ts [hunk 12]
```diff
@@ -185,7 +185,15 @@ export class EventDeployer {
     const deploymentExecution = await deploymentPreparation
     return async () => {
       try {
-        await deploymentExecution.execution()
+        const response: Timestamp | ErrorList = await deploymentExecution.execution()
+        if (typeof response != 'number') {
```
Comment (nchamo) [inline content/src/service/synchronization/EventDeployer.ts:189]: I'm not loving all the `typeof response != 'number'`. Let's create a helper/utils that given a deployment results, returns whether it was successful or not

File: content/test/unit/service/Service.spec.ts [hunk 13]
```diff
@@ -68,8 +68,14 @@ describe('Service', function () {
...
 
-    const timestamp: Timestamp = await service.deployEntity([entityFile, randomFile], entity.id, auditInfo, '')
-    const deltaMilliseconds = Date.now() - timestamp
+    const timestamp: Timestamp | ErrorList = await service.deployEntity(
+      [entityFile, randomFile],
+      entity.id,
+      auditInfo,
+      ''
+    )
+    expect(timestamp).toBeTruthy((a) => !(a instanceof Error))
```
Comment (nchamo) [inline content/test/unit/service/Service.spec.ts:77]: Why `Error`? Let's use the same assertion as before

File: content/test/integration/service/deployments/deployment-filters.spec.ts [hunk 14]
```diff
@@ -168,13 +168,15 @@ describe('Integration - Deployment Filters', () => {
...
       const newAuditInfo = { version: EntityVersion.V2, authChain: deployData.authChain, ...overrideAuditInfo }
-      const deploymentTimestamp = await service.deployEntity(
+      const deploymentTimestamp: Timestamp | ErrorList = await service.deployEntity(
         Array.from(deployData.files.values()),
         deployData.entityId,
         newAuditInfo,
         ''
       )
-      result.push(deploymentTimestamp)
+      if (typeof deploymentTimestamp === 'number') {
```
Comment (nchamo) [inline content/test/integration/service/deployments/deployment-filters.spec.ts:177]: If we use type guards correctly, then the `as Timestamp` shouldn't be necessary

File: content/test/integration/service/snapshots/SnapshotManager.spec.ts [hunk 15]
```diff
@@ -46,7 +46,8 @@ describe('Integration - Snapshot Manager', () => {
     // Assert snapshot was created
     const snapshotMetadata = snapshotManager.getSnapshotMetadata(EntityType.SCENE)
     expect(snapshotMetadata).toBeDefined()
-    expect(snapshotMetadata!.lastIncludedDeploymentTimestamp).toEqual(lastDeploymentTimestamp)
+    expect(lastDeploymentTimestamp).toBeTruthy((a) => !(a instanceof Error))
```
Comment (nchamo) [inline content/test/integration/service/snapshots/SnapshotManager.spec.ts:49]: Why `Error`? Also, let's create a helper (maybe in `E2EAssertions`?) that asserts that the deployment was successful

File: content/src/service/synchronization/EventDeployer.ts [hunk 16]
```diff
@@ -185,7 +185,15 @@ export class EventDeployer {
...
+          // The deployment failed, so we report it
+          await this.reportError(
+            deploymentExecution.metadata.deploymentEvent,
+            FailureReason.DEPLOYMENT_ERROR,
+            response.join()
```
Comment (nchamo) [inline content/src/service/synchronization/EventDeployer.ts:194]: Let's use the `.join('/n')`

File: content/src/service/synchronization/EventDeployer.ts [hunk 17]
```diff
@@ -170,7 +170,7 @@ export class EventDeployer {
 
   private buildDeploymentExecution(
     deploymentEvent: DeploymentWithAuditInfo,
-    execution: () => Promise<void>
+    execution: () => Promise<Timestamp | ErrorList>
```
Comment (nchamo) [inline content/src/service/synchronization/EventDeployer.ts:173]: Is this change necessary?

**Review by nchamo — COMMENTED** [review]
Great work!

Just come comments :)

File: content/src/service/synchronization/EventDeployer.ts (hunk 17 above)
Comment (agusaldasoro) [reply content/src/service/synchronization/EventDeployer.ts:173]: That was the way that I found to check whether the deployment was successful or not inside the `wrapDeployment` method, do you have any idea on how to avoid that?

File: content/test/integration/E2ETestUtils.ts (hunk 11 above)
Comment (agusaldasoro) [reply content/test/integration/E2ETestUtils.ts:126]: I wanted to reflect that the default value is error instead of a successful state to avoid false positives 🤔

File: content/test/integration/service/deployments/deployment-filters.spec.ts (hunk 14 above)
Comment (agusaldasoro) [reply content/test/integration/service/deployments/deployment-filters.spec.ts:177]: I couldn't make that work, if you could make the change proposal, that'd be great! 😄

Comment (nchamo) [reply content/test/integration/service/deployments/deployment-filters.spec.ts:177]: Let's try it after the `ErrorList` turns into an object

File: content/src/service/synchronization/EventDeployer.ts (hunk 17 above)
Comment (nchamo) [reply content/src/service/synchronization/EventDeployer.ts:173]: Oh, I see. Makes sense 😄

File: content/test/integration/E2ETestUtils.ts (hunk 11 above)
Comment (nchamo) [reply content/test/integration/E2ETestUtils.ts:126]: Oh, ok. Makes sense, good point!

File: content/src/service/ServiceImpl.ts [hunk 18]
```diff
@@ -147,17 +149,21 @@ export class ServiceImpl implements MetaverseContentService, ClusterDeploymentsS
...
     }
 
     // Update the current list of pointers being deployed
     entity.pointers.forEach((pointer) => pointersCurrentlyBeingDeployed.add(pointer))
     this.pointersBeingDeployed.set(entity.type, pointersCurrentlyBeingDeployed)
 
     try {
-      const { auditInfoComplete, wasEntityDeployed } = await repository.txIf(async (transaction) => {
+      const response:
+        | { auditInfoComplete: AuditInfo; wasEntityDeployed: boolean }
```
Comment (nchamo) [inline content/src/service/ServiceImpl.ts:165]: Is it ok that it starts with `|`?

File: content/test/integration/service/snapshots/SnapshotManager.spec.ts [hunk 19]
```diff
@@ -78,7 +80,8 @@ describe('Integration - Snapshot Manager', () => {
...
```
Comment (nchamo) [inline content/test/integration/service/snapshots/SnapshotManager.spec.ts:83]: Why `Error`? Shouldn't these 2 lines be replaced with `assertResultIsSuccessfulWithTimestamp`?

**Review by nchamo — APPROVED** [review]
Great work!

File: content/src/service/ServiceImpl.ts (hunk 18 above)
Comment (agusaldasoro) [reply content/src/service/ServiceImpl.ts:165]: It's added automatically by `prettier`
